File: PythonAnalizations/GigaChatService.py
import requests
import uuid
import time
import threading 
from typing import Optional, List, Dict, Any
import json
import re


class GigaChatService:
    def __init__(self):
        self.auth_key = "MDE5YTRhNzEtYWEyYy03MjM4LWExMjUtNTZmNTIwNDg1MTRhOjAzZTU1NDNkLWQ1MGQtNDVhMy1iYWU5LWE3ODkxY2Y4MzVkNA=="
        self.scope = "GIGACHAT_API_PERS"
        self.token: Optional[str] = None
        self.token_expiry: Optional[float] = None
        self.is_refreshing = False
        self.refresh_queue: List[Dict[str, Any]] = []
        self.lock = threading.Lock()

    def get_token(self, force_refresh: bool = False) -> str:
        """Получение токена доступа с кешированием и очередью запросов"""
        
        # Если токен есть и не истек, и не запрошено принудительное обновление
        if (self.token and self.token_expiry and 
            time.time() < self.token_expiry and not force_refresh):
            return self.token

        # Если уже идет обновление токена, добавляем запрос в очередь
        if self.is_refreshing:
            return self._wait_for_token_refresh()

        self.is_refreshing = True

        try:
            print("Requesting new GigaChat token...", {
                "clientId": "019a4a71-aa2c-7238-a125-56f52048514a",
                "rqUid": str(uuid.uuid4()),
            })

            url = "https://ngw.devices.sberbank.ru:9443/api/v2/oauth"
            headers = {
                "Content-Type": "application/x-www-form-urlencoded",
                "Accept": "application/json",
                "RqUID": str(uuid.uuid4()),
                "Authorization": f"Basic {self.auth_key}",
            }
            data = {"scope": self.scope}

            response = requests.post(
                url, 
                headers=headers, 
                data=data, 
                verify=False,
                timeout=10
            )
            response.raise_for_status()
            
            token_data = response.json()
            if not token_data.get("access_token"):
                raise ValueError("No access token in response")

            self.token = token_data["access_token"]
            # Токен действует 30 минут, устанавливаем expiry на 25 минут для запаса
            self.token_expiry = time.time() + 25 * 60

            print(
                "GigaChat token obtained successfully, expires at:",
                time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(self.token_expiry))
            )

            # Разрешаем все ожидающие запросы
            self._resolve_pending_requests(self.token)
            return self.token

        except Exception as error:
            print(f"GigaChat token error: {error}")
            # Отклоняем все ожидающие запросы
            self._reject_pending_requests(error)
            raise Exception(f"Failed to get GigaChat token: {error}")
        finally:
            self.is_refreshing = False

    def _wait_for_token_refresh(self) -> str:
        """Ожидание обновления токена другими запросами"""
        future = {"event": threading.Event(), "token": None, "error": None}
        with self.lock:
            self.refresh_queue.append(future)
        
        future["event"].wait()
        
        if future["error"]:
            raise future["error"]
        return future["token"]

    def _resolve_pending_requests(self, token: str):
        """Разрешение всех ожидающих запросов"""
        with self.lock:
            for future in self.refresh_queue:
                future["token"] = token
                future["event"].set()
            self.refresh_queue.clear()

    def _reject_pending_requests(self, error: Exception):
        """Отклонение всех ожидающих запросов"""
        with self.lock:
            for future in self.refresh_queue:
                future["error"] = error
                future["event"].set()
            self.refresh_queue.clear()

    def chat_completion(self, messages: List[Dict[str, str]], **kwargs) -> str:
        """Основной метод для отправки запросов к GigaChat API"""
        try:
            print("Sending request to GigaChat API...")

            token = self.get_token()
            url = "https://gigachat.devices.sberbank.ru/api/v1/chat/completions"
            headers = {
                "Content-Type": "application/json",
                "Accept": "application/json",
                "Authorization": f"Bearer {token}",
            }
            
            data = {
                "model": "GigaChat",
                "messages": messages,
                "stream": False,
                "repetition_penalty": 1,
                "temperature": 0.7,
                "max_tokens": 1000,
                **kwargs
            }

            response = requests.post(
                url,
                headers=headers,
                json=data,
                verify=False,
                timeout=30
            )
            response.raise_for_status()

            print("GigaChat API response received")
            result = response.json()
            return result["choices"][0]["message"]["content"]

        except requests.exceptions.HTTPError as error:
            if error.response.status_code == 401:
                print("Token expired, forcing refresh...")
                # Токен истек, принудительно обновляем
                self.token = None
                self.token_expiry = None
                raise Exception("Token expired, please retry")
            
            print(f"GigaChat API HTTP error: {error}")
            raise Exception(f"GigaChat API request failed: {error}")
        except Exception as error:
            print(f"GigaChat API error: {error}")
            raise Exception(f"GigaChat API request failed: {error}")

    def send_message(self, message: str, **kwargs) -> str:
        """Упрощенный метод для отправки одного сообщения"""
        messages = [{"role": "user", "content": message}]
        return self.chat_completion(messages, **kwargs)

    def refresh_token(self) -> str:
        """Принудительное обновление токена"""
        self.token = None
        self.token_expiry = None
        return self.get_token(force_refresh=True)

    def get_token_status(self) -> Dict[str, Any]:
        """Get token status"""
        return {
            "has_token": bool(self.token),
            "expires_at": time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(self.token_expiry)) 
                         if self.token_expiry else None,
            "is_expired": self.token_expiry is None or time.time() >= self.token_expiry,
            "time_until_expiry": self.token_expiry - time.time() 
                               if self.token_expiry else None,
        }



# Создаем глобальный экземпляр сервиса
giga_chat_service = GigaChatService()


# Упрощенный интерфейс для быстрого использования
def ask_gigachat(question: str) -> str:
    systemPromt = f"""
    
    """



    """Простая функция для быстрых запросов"""
    return giga_chat_service.send_message(question)




# Пример использования
if __name__ == "__main__":
    try:
        # Простой запрос
        response = ask_gigachat("Привет! Напиши короткое тестовое сообщение.")
        print("Ответ:", response)
        
        # Проверка статуса токена
        status = giga_chat_service.get_token_status()
        print("Статус токена:", status)
        
        
    except Exception as e:
        print(f"Ошибка: {e}")

File: AlgoVis.Server/Services/GigaChatService.cs
using AlgoVis.Server.Interfaces;
using System.Text;
using System.Text.Json;

namespace AlgoVis.Server.Services
{
    public class GigaChatService : IGigaChatService
    {
        private readonly string _authKey = "MDE5YTRhNzEtYWEyYy03MjM4LWExMjUtNTZmNTIwNDg1MTRhOjAzZTU1NDNkLWQ1MGQtNDVhMy1iYWU5LWE3ODkxY2Y4MzVkNA==";
        private readonly string _scope = "GIGACHAT_API_PERS";

        private string _token;
        private DateTime _tokenExpiry;
        private bool _isRefreshing;
        private readonly List<TaskCompletionSource<string>> _refreshQueue = new();
        private readonly SemaphoreSlim _lock = new SemaphoreSlim(1, 1);
        private readonly HttpClient _httpClient;

        public GigaChatService()
        {
            // Создаем HttpClient с отключенной проверкой SSL (аналог verify=False в Python)
            var handler = new HttpClientHandler
            {
                ServerCertificateCustomValidationCallback = (message, cert, chain, errors) => true
            };
            _httpClient = new HttpClient(handler);
        }

        public async Task<string> GetTokenAsync(bool forceRefresh = false)
        {
            // Если токен есть и не истек, и не запрошено принудительное обновление
            if (!string.IsNullOrEmpty(_token) &&
                _tokenExpiry > DateTime.Now &&
                !forceRefresh)
            {
                return _token;
            }

            // Если уже идет обновление токена, добавляем запрос в очередь
            if (_isRefreshing)
            {
                return await WaitForTokenRefreshAsync();
            }

            _isRefreshing = true;

            try
            {
                Console.WriteLine("Requesting new GigaChat token...");

                var url = "https://ngw.devices.sberbank.ru:9443/api/v2/oauth";
                var rqUid = Guid.NewGuid().ToString();

                var request = new HttpRequestMessage(HttpMethod.Post, url);
                request.Headers.Add("Accept", "application/json");
                request.Headers.Add("RqUID", rqUid);
                request.Headers.Add("Authorization", $"Basic {_authKey}");

                var content = new FormUrlEncodedContent(new[]
                {
                new KeyValuePair<string, string>("scope", _scope)
            });
                request.Content = content;

                var response = await _httpClient.SendAsync(request);
                response.EnsureSuccessStatusCode();

                var responseContent = await response.Content.ReadAsStringAsync();
                var tokenData = JsonSerializer.Deserialize<JsonElement>(responseContent);

                if (!tokenData.TryGetProperty("access_token", out var accessTokenProperty))
                {
                    throw new Exception("No access token in response");
                }

                _token = accessTokenProperty.GetString();
                // Токен действует 30 минут, устанавливаем expiry на 25 минут для запаса
                _tokenExpiry = DateTime.Now.AddMinutes(25);

                Console.WriteLine($"GigaChat token obtained successfully, expires at: {_tokenExpiry:yyyy-MM-dd HH:mm:ss}");

                // Разрешаем все ожидающие запросы
                ResolvePendingRequests(_token);
                return _token;
            }
            catch (Exception error)
            {
                Console.WriteLine($"GigaChat token error: {error}");
                // Отклоняем все ожидающие запросы
                RejectPendingRequests(error);
                throw new Exception($"Failed to get GigaChat token: {error.Message}");
            }
            finally
            {
                _isRefreshing = false;
            }
        }

        private async Task<string> WaitForTokenRefreshAsync()
        {
            var tcs = new TaskCompletionSource<string>();
            await _lock.WaitAsync();
            try
            {
                _refreshQueue.Add(tcs);
            }
            finally
            {
                _lock.Release();
            }

            return await tcs.Task;
        }

        private void ResolvePendingRequests(string token)
        {
            _lock.Wait();
            try
            {
                foreach (var tcs in _refreshQueue)
                {
                    tcs.SetResult(token);
                }
                _refreshQueue.Clear();
            }
            finally
            {
                _lock.Release();
            }
        }

        private void RejectPendingRequests(Exception error)
        {
            _lock.Wait();
            try
            {
                foreach (var tcs in _refreshQueue)
                {
                    tcs.SetException(error);
                }
                _refreshQueue.Clear();
            }
            finally
            {
                _lock.Release();
            }
        }

        public async Task<string> ChatCompletionAsync(List<Dictionary<string, string>> messages, Dictionary<string, object> additionalParams = null)
        {
            try
            {
                Console.WriteLine("Sending request to GigaChat API...");

                var token = await GetTokenAsync();
                var url = "https://gigachat.devices.sberbank.ru/api/v1/chat/completions";

                var requestData = new Dictionary<string, object>
                {
                    ["model"] = "GigaChat",
                    ["messages"] = messages,
                    ["stream"] = false,
                    ["repetition_penalty"] = 1,
                    ["temperature"] = 0.7,
                    ["max_tokens"] = 1000
                };

                // Добавляем дополнительные параметры если есть
                if (additionalParams != null)
                {
                    foreach (var param in additionalParams)
                    {
                        requestData[param.Key] = param.Value;
                    }
                }

                var request = new HttpRequestMessage(HttpMethod.Post, url);
                request.Headers.Add("Accept", "application/json");
                request.Headers.Add("Authorization", $"Bearer {token}");

                var jsonContent = JsonSerializer.Serialize(requestData);
                request.Content = new StringContent(jsonContent, Encoding.UTF8, "application/json");

                var response = await _httpClient.SendAsync(request);
                response.EnsureSuccessStatusCode();

                Console.WriteLine("GigaChat API response received");
                var responseContent = await response.Content.ReadAsStringAsync();
                var result = JsonSerializer.Deserialize<JsonElement>(responseContent);

                return result.GetProperty("choices")[0]
                            .GetProperty("message")
                            .GetProperty("content")
                            .GetString();
            }
            catch (HttpRequestException error) when (error.Message.Contains("401"))
            {
                Console.WriteLine("Token expired, forcing refresh...");
                // Токен истек, принудительно обновляем
                _token = null;
                _tokenExpiry = DateTime.MinValue;
                throw new Exception("Token expired, please retry");
            }
            catch (Exception error)
            {
                Console.WriteLine($"GigaChat API error: {error}");
                throw new Exception($"GigaChat API request failed: {error.Message}");
            }
        }

        public async Task<string> SendMessageAsync(string message, Dictionary<string, object> additionalParams = null)
        {
            var messages = new List<Dictionary<string, string>>
        {
            new Dictionary<string, string>
            {
                ["role"] = "user",
                ["content"] = message
            }
        };

            return await ChatCompletionAsync(messages, additionalParams);
        }

        public async Task<string> RefreshTokenAsync()
        {
            _token = null;
            _tokenExpiry = DateTime.MinValue;
            return await GetTokenAsync(true);
        }

        public string GeneratePromt(string code)
        {
            string content = File.ReadAllText("Instruction.txt");

            string finalpromt = $"Есть инструкция по языку ява : \"{content}\"," +
                $"мне нужно чтобы ты переписал вот этот код: {code} в язык ЯВА и вернул ответ в виде JSON c алгоритмом как в инструкции и без лишних слов(только Json алгоритма)";

            return finalpromt;
        }

        public Dictionary<string, object> GetTokenStatus()
        {
            return new Dictionary<string, object>
            {
                ["has_token"] = !string.IsNullOrEmpty(_token),
                ["expires_at"] = _tokenExpiry.ToString("yyyy-MM-dd HH:mm:ss"),
                ["is_expired"] = string.IsNullOrEmpty(_token) || DateTime.Now >= _tokenExpiry,
                ["time_until_expiry"] = string.IsNullOrEmpty(_token) ? null : (_tokenExpiry - DateTime.Now).TotalSeconds
            };
        }
    }
}
